### ISSUE TITLE
`update` gets slower with the number of modules in project

### ISSUE THREAD

**Issue by jtjeferreira** [sbt/sbt#5508]
## steps

Checkout https://github.com/jtjeferreira/sbt-multi-module-sample which is a sample of a multi module project with a dependency to `spark` (spark was chosen because it has a lot of transitive dependencies).

Run `sbt update`. After everything is downloaded for the first time, run `sbt update` again so no downloads are made again. In my dell xps 13 laptop with encrypted disk this takes 35s. 

If the project just contains a single module, `update` just takes 5s (see https://github.com/jtjeferreira/sbt-multi-module-sample/commit/4f96638ec5802dac22368c9905975578ba481f27)

## problem

It seems that sbt/coursier is validating the checksum of the artifacts once per module, which is unnecessary...

## expectation

I would expect that `sbt update` duration should just increase with the number of artifacts it needs to check, not with the number of modules... 

## notes

I have run Java Flight Recorder and can report the following:
* A lot of `coursier-pool-n-thread-m` are created. Where n is between [2, 162], and m is between [1, 5] and in some cases [1,6]. That means around 1k threads were created
* `coursier.cache.FileCache#validateChecksum` seems to be the hotpath

**Comment by jtjeferreira** [comment]
update: if ivy and cachedResolution is used instead, it takes around 19s (vs 35s with coursier...)

**Comment by eed3si9n** [comment]
@jtjeferreira Thanks for the report. Given that you've attributed this behavior to Cousier, could you report this to coursier/coursier as well please?

**Comment by jtjeferreira** [comment]
Hi @eed3si9n, sure I can open, but I am not sure the problem is in sbt or coursier. An possible solution would be that sbt calls coursier differently; or that sbt implements cachedResolution for coursier and calls coursier less often?

Or do you have the feeling this is definitely something to be fixed on coursier side?

**Comment by eed3si9n** [comment]
I could be wrong, but I am guessing it's on Coursier side since I am not aware of sbt calling `validateChecksum`.

**Comment by jtjeferreira** [comment]
> I could be wrong, but I am guessing it's on Coursier side since I am not aware of sbt calling `validateChecksum`.

Yes, thats true, but let me elaborate better my assumptions and also what I investigated so far:
* `coursier.cache.FileCache#validateChecksum` is the hotpath, but this is because 
sbt is calling coursier (`coursier.Artifacts.either()`) for every module, and then `coursier.cache.FileCache#file` and thats where `validateChecksum` will be  called...
* coursier already has a `Cache` concept, but the default implementation, which is used by `sbt` is `FileCache` which basically represents `~/.coursier/cache` and everytime a file is requested, checksum validation happens...
* on sbt side `lmcoursier.internal.Lock#lock` also looks suspicious, and according to JFR/JMC there is a lot of blocking going on

![image](https://user-images.githubusercontent.com/943051/79780342-abf8c700-8333-11ea-8822-c5eda4267733.png)

**Comment by jtjeferreira** [comment]
So I created a ticket upstream https://github.com/coursier/coursier/issues/1675 

Also if this is fixed in coursier it will also benefit other projects than sbt

**Comment by alexarchambault** [comment]
@jtjeferreira Did you try disabling checksums via `csrConfiguration`? Something like
```scala
csrConfiguration := csrConfiguration.value.withChecksums(Vector(None))
```
should do I think. (Maybe this needs to be applied to all projects?)

I'm not that surprised checksums are a problem, although you're getting quite large numbers here… I usually disable checksums when I benchmark coursier.

**Comment by jtjeferreira** [comment]
Hi @alexarchambault,

thanks for that tip, I didn't notice that configuration before.

However, it only reduced the time from 35s to 29s (https://github.com/jtjeferreira/sbt-multi-module-sample/commit/829e33f5ff3e15485299f07517b7a78abbccbaf1), and in JFR I still see `validateChecksum` in the hotpath

![image](https://user-images.githubusercontent.com/943051/79797312-f0459080-834e-11ea-8664-90061335b545.png)

**Comment by alexarchambault** [comment]
Did you try putting the setting above in all projects?

**Comment by jtjeferreira** [comment]
Indeed I was not setting it for all modules...

After setting it for all modules (https://github.com/jtjeferreira/sbt-multi-module-sample/commit/db9fd1945bce5b80ea6406943bbc9984b42947de) the time was reduced to 26s, and `validateChecksum` is no longer present...

![image](https://user-images.githubusercontent.com/943051/79812006-7c67b000-836f-11ea-9494-ad07c9ea243e.png)

I see a few collection optimizations that can be done, but I am not sure how much it will impact...

**Comment by jtjeferreira** [comment]
Hi @eed3si9n 

not wanting to start any flame war or discussion about which tool is better, I tried to compare a similar project structure in mill (which also uses coursier)

The sample project is in https://github.com/jtjeferreira/mill-multi-module-sample and used `mill show __.compileClasspath` as the similar command to `update`. The result is that mill is very fast (5s). My understanding is that mill uses much more aggressive caching, but cant this caching be also done in `sbt`?

**Comment by eatkins** [comment]
@jtjeferreira does sbt take 25 seconds to run update every time? Or if you have an active sbt session does `update` run significantly faster the second time? The sbt update report data structure is cached to disk but the deserialization is slow (https://github.com/sbt/sbt/issues/4883) for large dependency graphs. When sbt first runs `update` it has to deserialized the cached report from disk but the report should be persisted in memory which usually makes subsequent calls to update much faster in my experience. If the second call to update remains slow, then it's likely that the in-memory update report is being invalidated for whatever reason.

Mill and sbt have similar caching capabilities but sbt's have been added in a somewhat ad-hoc fashion and are not universally applied for all tasks whereas in mill task result caching is a core part of the design. Any hotspots in sbt that are not using caching could be updated using the apis in [Track File Inputs and Outputs](https://www.scala-sbt.org/1.x/docs/Howto-Track-File-Inputs-and-Outputs.html) and [Previous value](https://www.scala-sbt.org/1.x/docs/Caching.html#Previous+value). The `update` task is specifically a bit tricky to cache because the report can have an expiration time. I don't know if/how mill handles that since all that I've read about its caching approach was based on tracking source files but expiration definitely makes `update` more complicated in sbt.

**Comment by eed3si9n** [comment]
Yea. As I reported in https://github.com/sbt/sbt/issues/4883 `UpdateReport` persistence is slow. I am totally open to specializing the `update` caching to non-JSON based ways, or try different strategies to speed things up.

**Comment by jtjeferreira** [comment]
> @jtjeferreira does sbt take 25 seconds to run update every time? Or if you have an active sbt session does `update` run significantly faster the second time? The sbt update report data structure is cached to disk but the deserialization is slow (#4883) for large dependency graphs. When sbt first runs `update` it has to deserialized the cached report from disk but the report should be persisted in memory which usually makes subsequent calls to update much faster in my experience. If the second call to update remains slow, then it's likely that the in-memory update report is being invalidated for whatever reason.
> 

I know subsequent calls are faster, but what I am trying to improve is the "boot time" of sbt in a project with a lot of modules...

So focusing on the first call of `update`, and according to your description, why is sbt calling coursier? Isn't the report already in disk? From the flamegraph I collected the majority of samples are coursier related:

![image](https://user-images.githubusercontent.com/943051/80477980-b83cdf80-8944-11ea-9d0d-af3cff5e1e0e.png)

(the brighter yellow is the result of searching for "lm.coursier")

**Comment by eatkins** [comment]
> I know subsequent calls are faster, but what I am trying to improve is the "boot time" of sbt in a project with a lot of modules...

A worthy goal. sbt hasn't always historically prioritized boot time.

I am almost certain that the lm.coursier stuff is just a red herring. Most of what I see in the flamegraph is [sjson-new](https://github.com/eed3si9n/sjson-new) deserialization. The most likely way to make that faster is to improve sjson-new or to simplify the serialized update report so that it can be deserialized more quickly.

**Comment by jtjeferreira** [comment]
> I am almost certain that the lm.coursier stuff is just a red herring. Most of what I see in the flamegraph is [sjson-new](https://github.com/eed3si9n/sjson-new) deserialization. The most likely way to make that faster is to improve sjson-new or to simplify the serialized update report so that it can be deserialized more quickly.

I am also doing other PRs to improve the sjson-new serialization, but from looking at the flamegraph only a small part is sjson serialization:

![image](https://user-images.githubusercontent.com/943051/80537969-d5050180-899c-11ea-9105-b88d22506030.png)

However something in this serialized update report is not clear to me yet: if the report is serialized in disk, then calling coursier again should not be necessary right? do you know the location of this report? and where this report is checked?

Thanks in advance...

**Comment by eatkins** [comment]
You're correct that deserialization is not the only bottleneck and that coursier is also doing something slow. I changed the updateTask0 function in Defaults.scala to by a dynamic task so that the first few lines look like:
```scala
  private def updateTask0(
      cacheLabel: String,
      includeCallers: Boolean,
      includeDetails: Boolean
  ): Initialize[Task[UpdateReport]] = Def.taskDyn {
    import sbt.librarymanagement.LibraryManagementCodec._
    update.previous match {
      case Some(p) => Def.task(p)
      case _ =>
        Def.task {
          val s = streams.value
          val cacheDirectory = crossTarget.value / cacheLabel / updateCacheName.value
          ...
}
```

After that change, it is still very slow to update the first time but subsequent runs are faster. In fact, update was instantaneous when run the second time within a session whereas before it was still quite slow the second time.

With your sample project at the top of the issue, I got:
```
> time sbt -Dsbt.version=1.4.0-SNAPSHOT update
...
sbt -Dsbt.version=1.4.0-SNAPSHOT update  63.28s user 3.53s system 452% cpu 14.774 total
```
This was about 4.5 seconds slower than `sbt exit` which means that deserialization was taking about 4.5 seconds but the performance penalty of the coursier stuff is much higher. I have not seen the coursier overhead in the past but I had seen the slow deserialization which is why I had assumed that was the problem.

**Comment by neko-kai** [comment]
My project has 155 modules, this problem is very relevant to me – the update command always takes mor than 2 minutes, it's so bad that it takes way more time to import the project into Intellij IDEA than it takes to `clean; test:compile` it.

In case it can help, we may make a reproduction of that.

**Comment by eatkins** [comment]
Just to follow up, sbt 1.4.0 doesn't really do anything to fix the underlying issue with update being slow. I agree with @jtjeferreira that update time should theoretically be proportional to the number of artifacts that are resolved rather than the number of sub projects. That being said, I do find that once sbt has run compile once that the overhead of update is very small for subsequent compiles. My hope is that with sbt 1.4.0 that the impact of this issue won't be felt as badly since users can use sbtn or leave the sbt console open in their terminal. Prior to 1.4.0-RC1, sbt was leaking memory with every task evaluation so in a large multi module project, the sbt server would end up running low on memory fairly quickly which really degraded performance or even caused an OOM. I can now run 1000 no-op compiles with no degraded performance in https://github.com/jtjeferreira/sbt-multi-module-sample whereas with 1.4.0-M2 and older, the performance degrades noticeably after a few dozen no-op compiles.

**Comment by jtjeferreira** [comment]
FYI `COURSIER_PROGRESS=false` improves things dramatically https://github.com/coursier/coursier/issues/1720#issuecomment-731168494

> @oyvindberg fantastic work, this reduced update from 30s to 10s in my work project. I will update the sample project with new results and try to find where other bottlenecks might be...

**Comment by SethTisue** [comment]
is it possible to set `COURSIER_PROGRESS=false` when it's IntellIJ running the `update` ?

**Comment by oyvindberg** [comment]
I haven't managed to yet @SethTisue . 

Just to continue iterating slowly, the easiest fix would be to teach coursier to understand JVM properties as well, as those can more easily be configured, also in Intellij.

After that we need to find a path forward to bring this performance improvement to all users. I have expressed before that I'd rather remove the coursier progress bars entirely (and with them the need to block threads). Then we could work with sbt to add support for rendering progress bars for tasks and reimplement it there. See https://github.com/sbt/sbt/issues/5627 , which I unfortunately didn't have time to follow up.

**Comment by jtjeferreira** [comment]
Hi!

I tried the sample project with more recent versions of sbt (1.5.8 vs 1.7.3) in https://github.com/jtjeferreira/sbt-multi-module-sample/tree/d1d3fc8700787b3ad5824d34b6f722fff96ea58c and these were the results

| run                                            | time | update task |
|------------------------------------------------|------|-------------|
| 60 modules (sbt 1.5.8)                         | 36s  | 26s         |
| 60 modules (sbt 1.5.8) COURSIER_PROGRESS=false | 25s  | 15s         |
| 60 modules (sbt 1.7.3)                         | 31s  | 22s         |
| 60 modules (sbt 1.7.3) COURSIER_PROGRESS=false | 18s  | 9s          |

* `COURSIER_PROGRESS=false` still improves performance a lot (26s vs 15s in sbt 1.5.8 or 22s vs 9s in sbt 1.7.3)
* sbt 1.7.3 also improves performance (15s vs 9s)

**Comment by oyvindberg** [comment]
Time to make it the default for sure

**Comment by danicheg** [comment]
Hey there. First, thanks for this ad-hoc workaround that still improves 2-3x time of `sbt update` for a project with hundreds of sbt modules. I'm wondering about the reasons this is not opted-in by default 🤔

**Comment by oyvindberg** [comment]
Beats me. 

This fix should *absolutely* be part of sbt 2. I'm sure the performance improvement would be a stronger reason to update than the Scala 3 update for many. 

Just last week I watched a coworker hit "reimport build" in intellij and wait *4 minutes* for it to finish.

**Comment by danicheg** [comment]
There's nothing to be proud of, but we (at $job) are at the point of "Oh, wow, 4 minutes, not bad".

**Comment by vbabenkoru** [comment]
Yeah I wish ours was 4 minutes.. it is closer to 17-20 minutes to run update.
